### PR TITLE
support sled transaction for metasrv

### LINKED ISSUE #2422: [metasrv] Support sled tree level transaction
**Summary**

Support sled tree level transaction for metasrv

**Tasks**
Implementation path:

- [ ] Identify all the codes that need transaction support.
- [x] Define the necessary transaction API and data structures.
- [ ] [[metasrv] make sled support range read](#3309 )
- [ ] Impl a common third-party kv trait that can integrate different kv db.
- [ ] Integrate Rocksdb
- [ ] Test
    - [x] Test when `StateMachine::apply` supports transaction (#3102 )
	- [ ] Test after support sled range read
	- [ ] Change test after we abstract a common kv trait
	- [ ] Test after integrate Rocksdb

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

MetaSrv operations support transaction based on Sled

## Changelog

- Improvement

## Related Issues

#2422 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/meta/raft-store/src/state_machine/sm.rs [hunk 1]
```diff
@@ -248,9 +250,15 @@ impl StateMachine {
...
         let sm_meta = self.sm_meta();
-        sm_meta
-            .insert(&LastApplied, &StateMachineMetaValue::LogId(*log_id))
-            .await?;
+        sm_meta.txn(move |t| {
+            t.insert(&LastApplied, &StateMachineMetaValue::LogId(*log_id))
+                .await?;
+            t.txn_tree.flush();
+            Ok(())
+        });
```
Comment (drmingdrmer) [inline common/meta/raft-store/src/state_machine/sm.rs:258]: A transaction may include more than one KeySpace, thus here it should be something like:
```rust
let txn_tree = TxnTree::new(self.tree);
txn_tree.txn(|t| {
    t.sm_meta().insert(k,v)
})
```

File: common/exception/src/exception.rs [hunk 2]
```diff
@@ -45,6 +51,139 @@ impl ToString for ErrorCodeBacktrace {
...
+        match e {
+            UnabortableTransactionError::Storage(e) => Self::Storage(e.into()),
+            UnabortableTransactionError::Conflict => Self::Conflict,
+        }
+    }
+}
+
+/// An error returned from the transaction closure
+#[derive(Debug)]
+pub enum SledConflictableTransactionError<T> {
```
Comment (drmingdrmer) [inline common/exception/src/exception.rs:138]: If it is an error returned from a tx closure, it better be a internal error type.
common/exception is a container for error types that are used by more than one crates.

File: common/exception/src/exception.rs [hunk 3]
```diff
@@ -45,6 +51,139 @@ impl ToString for ErrorCodeBacktrace {
...
     }
 }
+/// Put Sled's errors wrapper here to avoid circular references
+/// Error returned by a Sled transaction
+#[derive(Debug)]
+pub enum SledTransactionError<T> {
+    Abort(T),
+    /// A storage related error
+    Storage(io::Error),
+}
```
Comment (drmingdrmer) [inline common/exception/src/exception.rs:61]: `#[derive(this_error::Error)]` would save a lot lines of code for you 🤔

Comment (BohuTANG) [reply common/exception/src/exception.rs:61]: Must flush a :rocket:  to you @drmingdrmer

**Comment by ariesdevil** [comment]
@drmingdrmer I update some codes, PTAL again. If the current design is acceptable, I'll go on.
PS: please ignore errors impl in `exception.rs` and test errors for now:)

File: common/meta/sled-store/src/sled_tree.rs [hunk 4]
```diff
@@ -467,12 +479,72 @@ impl SledTree {
...
+    where
+        KV: SledKeySpace,
+    {
+        let k = KV::serialize_key(key).unwrap();
+        let v = KV::serialize_value(value).unwrap();
+
+        let prev = self
+            .txn_tree
+            .insert(k, v)
+            .map_err(SledUnabortableTransactionError::from)?;
```
Comment (drmingdrmer) [inline common/meta/sled-store/src/sled_tree.rs:520]: Why do we need to map the error here?

The caller should still be in a sled txn context, and the sled::transaction expects a `ConflictableTransactionResult` error right?

File: common/meta/sled-store/src/sled_tree.rs [hunk 5]
```diff
@@ -84,6 +88,16 @@ impl SledTree {
         }
     }
 
+    pub fn txn<T, E>(
+        &self,
+        f: impl Fn(TransactionSledTree<'_>) -> Result<T, SledConflictableTransactionError<E>>,
+    ) -> Result<T, SledTransactionError<E>> {
```
Comment (drmingdrmer) [inline common/meta/sled-store/src/sled_tree.rs:94]: Why not just let `f` return `ConflictableTransactionResult`?

File: common/meta/sled-store/src/sled_tree.rs [hunk 6]
```diff
@@ -97,6 +117,53 @@ impl SledTree {
         Ok(got)
     }
 
+    /// Retrieve the value of key.
+    pub fn get<KV: SledKeySpace>(&self, key: &KV::K) -> common_exception::Result<Option<KV::V>> {
+        let got = self
```
Comment (drmingdrmer) [inline common/meta/sled-store/src/sled_tree.rs:122]: If it is possible, do not move codes in a PR that adds new feature.

Hard to see if the moved code is changed or not... :(((

**Comment by drmingdrmer** [comment]
@ariesdevil 
What's the status of this PR? A draft to show the general idea or a complete small step ready to merge?

**Comment by ariesdevil** [comment]
> @ariesdevil What's the status of this PR? A draft to show the general idea or a complete small step ready to merge?

If this design is acceptable, I want to merge it first, then I'll go on to modify other codes that need transaction supported.

**Comment by drmingdrmer** [comment]
> If this design is acceptable, I want to merge it first, then I'll go on to modify other codes that need transaction supported.

Quite good.
Just do it.

**Comment by BohuTANG** [comment]
Good job~

**Pull request merged**